### PR TITLE
[TS] LPS-188014 Cluster

### PR DESCRIPTION
**Pull request by yuhai**
@tinatian, @dantewang,

The previous pr is https://github.com/liferay-core-infra/liferay-portal/pull/4826.

In order to solve SF issue, I asked help with Alan to add the [SF](https://github.com/liferay-core-infra/liferay-portal/commit/8744fc41b5ae5f1c14430ab7a2a3f4ed398422c7).

### CODE REVIEW

**Comment by yuhai** [comment]
ci:test:clustering

**Comment by zhangyan0701** [comment]
ci:stop

**Comment by yuhai** [comment]
ci:test:sf

**Comment by yuhai** [comment]
ci:test:relevant

**Comment by yuhai** [comment]
ci:test:clustering

**Comment by Kvale1733** [comment]
@yuhai No regressions found. Testing passed following ticket steps.

**Comment by tinatian** [comment]
@yuhai Does this affect start up performance?

**Comment by dantewang** [comment]
@GinsonRen @111Anna Could you run startup diff on this PR?

**Comment by yuhai** [comment]
> @yuhai Does this affect start up performance?

@tinatian， I just tested startup DXP portal in the cluster environment without the issue(empty jgroupsping data) on my side (-Xms2560m -Xmx2560m -XX:MaxNewSize=1536m -XX:MaxMetaspaceSize=768m -XX:MetaspaceSize=768m -XX:NewSize=1536m).
**hot startup with the fix.**
1) 23302
2) 20648
3) 22401
4) 21110
5) 20817
**Average: 21655.6**

**hot startup without the fix.**
1) 20064
2) 20223
3) 20484
4) 19850
5) 21389
**Average: 20402**

It seems affected performance. I just asked @111Anna help to run the test in our startup test environment.

**Comment by dantewang** [comment]
@yuhai Thanks, let's wait for Anna's result.

**Comment by 111Anna** [comment]
@tinatian @dantewang @yuhai @GinsonRen    I have tested the PR on test machine, it seems to have no effect on startup performance.

 Here are test results:

CE:

Portal Version |GitId | Average Startup (ms) | % | Average Shutdown (ms) | % | Open File |Thread Create |Open Prepared Statement |Open SQL|Open service tracker|Open bundle tracker 
-- | -- | -- | -- | -- | -- | -- | -- | -- | --|--|--
Baseline | 032c27b5 |18719.6 | - | 2645.6 | - | 1456 | 94 | 120 | 837|5345|33
Baseline + PR-4899 | - | 18880.3 | -0.85% | 2589.3 | 2.1% |  1456|  94| 120 | 836|5347|33


DXP:

Portal Version | GitId | Average Startup (ms) | % | Average Shutdown (ms) | % | Open File |Thread Create |Open Prepared Statement |Open SQL|Open service tracker|Open bundle tracker 
-- | -- | -- | -- | -- | -- | -- | -- | -- | --|--|--
Baseline | 032c27b5 |19861.6 | - | 3026.3 | - | 1559 | 94 | 128 | 905|5723|33
Baseline + PR-4899 | - | 19757.3 | 0.52% | 3024.6 | 0.05% | 1559 |97 | 126 |904 |5723|33

File: modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java [hunk 1]
```diff
@@ -135,10 +144,52 @@ public void complete(Future<Void> future) {
...
+					countDownLatch.countDown();
+				};
+
+				bundleContext.addServiceListener(
+					serviceListener, "(&(objectClass=" + componentName + "))");
+
+				_componentContext.enableComponent(componentName);
+
+				while (true) {
+					if (countDownLatch.await(1, TimeUnit.MINUTES)) {
```
Comment (dantewang) [inline modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java:175]: @yuhai Is it necessary to wait for the `countDownLatch` every 1 min in a infinite loop? We should be able to just use the `await()` method, which will block the current thread until `countDown()`.

Comment (yuhai) [reply modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java]: @dantewang, I just push force the SF commit to change this.

**Comment by yuhai** [comment]
ci:test:sf

**Review by dantewang — APPROVED** [review]

**Comment by dantewang** [comment]
ci:test:startup

**Comment by dantewang** [comment]
ci:test

**Comment by catarineaguiar** [comment]
no regressions found

**Comment by catarineaguiar** [comment]
SF issue is not related

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/JavaOSGiReferenceCheck.java [hunk 2]
```diff
@@ -106,7 +106,9 @@ private void _checkMissingReference(String fileName, String content) {
 		while (matcher.find()) {
 			String serviceUtilClassName = matcher.group(2);
 
-			if (serviceUtilClassName.equals("IdentifiableOSGiServiceUtil")) {
+			if (serviceUtilClassName.equals("IdentifiableOSGiServiceUtil") ||
+				serviceUtilClassName.equals("SystemExecutorServiceUtil")) {
+
```
Comment (tinatian) [inline modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/JavaOSGiReferenceCheck.java:111]: @dantewang @yuhai Why this is needed?

File: modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java [hunk 3]
```diff
@@ -171,15 +171,11 @@ private void _registerDependencyManagerSyncGatekeeperCallable() {
...
-						_componentContext.disableComponent(componentName);
+				countDownLatch.await();
 
-						bundleContext.removeServiceListener(serviceListener);
+				_componentContext.disableComponent(componentName);
 
-						break;
-					}
-				}
+				bundleContext.removeServiceListener(serviceListener);
```
Comment (tinatian) [inline modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java:178]: @yuhai Please do not have a separate commit to modify what you just added, modify the original commit directly.

**Comment by yuhai** [comment]
@tinatian, our fix can expose one SF bug. I just separate the SF fix to send. The SF fix pr is https://github.com/ling-alan-huang/liferay-portal/pull/1991. So the current pr is based on that. I merge simplify fix to the original commit and adjust the commits order. No other changes and needn't to trigger ci test.

**Comment by tinatian** [comment]
See https://github.com/liferay-core-infra/liferay-portal/pull/4989

**Comment by dantewang** [comment]
Closing in favor of #4989

File: modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java [hunk 4]
```diff
@@ -135,10 +144,48 @@ public void complete(Future<Void> future) {
...
+
+				BundleContext bundleContext =
+					_componentContext.getBundleContext();
+
+				ServiceListener serviceListener = serviceEvent -> {
+					if (serviceEvent.getType() != ServiceEvent.REGISTERED) {
+						return;
+					}
+
+					bundleContext.getService(
```
Comment (tinatian) [inline modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java:163]: @yuhai @dantewang Why this getService is needed?

Comment (yuhai) [reply modules/apps/static/portal-dependency-manager-component-executor-factory/portal-dependency-manager-component-executor-factory/src/main/java/com/liferay/portal/dependency/manager/component/executor/factory/internal/DependencyManagerSyncImpl.java:163]: @tinatian, getService() means DependencyManagerSyncGatekeeper service activate() finished. And then runs the following logic.
